src/Spanner/Session/CacheSessionPool.php

```diff
@@ -0,0 +1,548 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Spanner\Session;
+
+use Google\Cloud\Core\Lock\LockInterface;
+use Google\Cloud\Core\Lock\SymfonyLockAdapter;
+use Google\Cloud\Spanner\Database;
+use Psr\Cache\CacheItemPoolInterface;
+use Symfony\Component\Lock\Factory;
+use Symfony\Component\Lock\Store\FlockStore;
+
+/**
+ * This session pool implementation accepts a PSR-6 compatible cache
+ * implementation and utilizies it to store sessions between requests.
+ *
+ * Please note that if you configure a high minimum session value the first
+ * request and any after a period of inactivity greater than an hour (the point
+ * at which sessions will expire) will have an increased amount of latency. This
+ * is due to the pool attempting to create as many sessions as needed to fill
+ * itself to match the minimum value.
+ *
+ * For this reason, it is highly recommended to configure a script to make an
+ * initial request to warm up the pool and manage subsequent requests during
+ * off-peak hours to keep the pool active.
+ *
+ * Please note: While required for the session pool, a PSR-6 implementation is
+ * not included in this library. It will be neccesary to include a separate
+ * dependency to fulfill this requirement.
+ *
+ * Example:
+ * ```
+ * use Google\Cloud\Spanner\SpannerClient;
+ * use Google\Cloud\Spanner\Session\CacheSessionPool;
+ * use Symfony\Component\Cache\Adapter\FilesystemAdapter;
+ *
+ * $spanner = new SpannerClient();
+ * $cache = new FilesystemAdapter();
+ * $sessionPool = new CacheSessionPool($cache);
+ *
+ * $database = $spanner->connect('my-instance', 'my-database', [
+ *     'sessionPool' => $sessionPool
+ * ]);
+ * ```
+ */
+class CacheSessionPool implements SessionPoolInterface
+{
+    const CACHE_KEY_TEMPLATE = 'cache-session-pool.%s.%s';
+
+    /**
+     * @var array
+     */
+    private static $defaultConfig = [
+        'maxSessions' => PHP_INT_MAX,
+        'minSessions' => 1,
+        'shouldWaitForSession' => true,
+        'maxCyclesToWaitForSession' => 30,
+        'sleepIntervalSeconds' => .5
+    ];
+
+    /**
+     * @var CacheItemPoolInterface
+     */
+    private $cacheItemPool;
+
+    /**
+     * @var string
+     */
+    private $cacheKey;
+
+    /**
+     * @var array
+     */
+    private $config;
+
+    /**
+     * @var Database
+     */
+    private $database;
+
+    /**
+     * @param CacheItemPoolInterface $cacheItemPool A PSR-6 compatible cache
+     *        implementation used to store the session data.
+     * @param array $config [optional] {
+     *     Configuration Options.
+     *
+     *     @type int $maxSessions The maximum number of sessions to store in the
```

Comment:
I thought a bit more about this and the related issue of close. We have already seen cases where people have hit max session limits on the backend and locked themselves out for an hour so we should try to avoid that as much as possible. I think we should add some way to downsize the pool. Otherwise a sudden increase in traffic will create lot of sessions which will never go away.  But we should not downsize to min sessions otherwise with the default value of 1 for that we will just keep on deleting sessions. Ideally we would keep track of how many sessions are needed to support the current load and downsize to that. I am not sure what is the best way to do this in PHP. Maybe we can add a downsize method and leave it to the user to figure out how and when to call it.